### ISSUE TITLE
Legacy Update fails to install on Windows 8.1

### ISSUE THREAD

**Issue by GriffintheFolf** [kirb/LegacyUpdate#104]
On Windows 8.1 32-bit, when trying to install LU, one of the prerequisites, "Windows 8.1 Update 1 (KB2959977)" fails to install with a nonsensical error code.

I do not believe this issue is relating to inadequate system resources; the system has 2 GB of RAM and more than 100 GB of free disk space.

Although unconfirmed, other people I have asked said they have experienced similar issues.
![image](https://user-images.githubusercontent.com/26304783/214384424-5c9892f9-9b1a-4fe9-b903-d3c7b2cf0926.png)

**Comment by kirb** [comment]
That error matches up to [80240017](https://legacyupdate.net/errorcodes#error_80240017), "Operation was not performed because there are no applicable updates". I have a feeling 32-bit is the core of the issue. I'll need to download 32-bit 8.1 and give it a test.

Can you click Start &rarr; type in "update history" &rarr; take a screenshot of that list?

Since you have Firefox installed, you could also directly download 8.1 Update 1 from this link: https://www.microsoft.com/en-us/download/details.aspx?id=42327. See if you can make it work by installing manually. They need to be installed in this exact order:

* ClearCompressionFlag.exe
* Windows8.1-KB2919355-x86.msu
* Windows8.1-KB2932046-x86.msu
* Windows8.1-KB2959977-x86.msu
* Windows8.1-KB2937592-x86.msu
* Windows8.1-KB2938439-x86.msu
* Windows8.1-KB2934018-x86.msu

You won't see anything when you run ClearCompressionFlag.exe, just double click and give it a few seconds. KB2919355 and KB2932046 in your case will probably say already installed, but run them anyway just in case. When you get to KB2959977, it'll probably display the error string I mentioned before. If so, just continue to the next update anyway, see if they'll install without issue. No need to restart till the last update is done. If you get the Update 1 features after reboot (power button on Start screen, Windows Store app in your taskbar), it worked.

**Comment by GriffintheFolf** [comment]
> If you get the Update 1 features after reboot (power button on Start screen, Windows Store app in your taskbar), it worked.

I actually had the Store icon in my taskbar on a fresh installation, but I unpinned it since I didn't envision that I would need to use it, nor did I think it was an indication of what updates were installed. In that regard, here's the list from before manually installing the updates:
![image](https://user-images.githubusercontent.com/26304783/214474100-cabdf9c7-a42b-47c1-9895-1ba026f0a10e.png)


As for installing the said updates in the order that you specified, every one of them said it was already installed, except for KB2959977 erroring with "The update is not applicable to your computer" like you predicted.
![image](https://user-images.githubusercontent.com/26304783/214474195-5a6da209-4a18-4eb6-9bb1-24a18b05105e.png)

**Comment by stdin82** [comment]
ClearCompressionFlag.exe should not be executed if KB2919355 is detected as installed

KB2959977 is long gone superseded update (for WUA)
you can safely replace it with KB3044374

**Comment by kirb** [comment]
Thanks, I'll have to go back over it again but there was a whole thing about how this WUA update had to be used due to a catch-22 with the prerequisite criteria on later WUA updates. If I recall right, KB2959977 was the last one to *not* have that issue, so despite being superseded, it needs to be installed to unblock later WUA updates from being installed. Anyway, I think it makes sense for us to treat that error code as being successful - just ignore and keep going, because the way we check for the update is naive and doesn't handle supersedence (by looking for the exact servicing package KB and revision being installed). Only reason I didn't already do that is because I wanted to test what happens when installing the updates on Embedded 7, 8, 8.1.

**Comment by GriffintheFolf** [comment]
Upon further examination, the ISO I used to install Windows was Update 3 (build 9600), which was downloaded from Rufus since the official Microsoft page doesn't seem to work anymore. 

I am preparing another system with the same download (except in 64-bit) in case any further platform-specific issues arise.

**Comment by kirb** [comment]
Aha. I was thinking so. In which case, none of the updates should be needed, since they're already baked in. I'll download this as well so I can make sure they don't get installed on this iso.

**Comment by Niklas20114552** [comment]
I am having the exact issue with legacyupdate. Is there already a fix, that work's without reinstalling the system?
I am using Windows 8.1 Pro Build 9600 too.

**Comment by renodr** [comment]
I've been looking into this a little bit, and it seems that on a fresh Windows 8.1 Update 3 install (downloaded from the ISO on Microsoft's website as a test), four out of five of the KBs are installed - the missing one is KB2959977.

See [https://support.microsoft.com/en-us/topic/windows-update-client-does-not-scan-against-wsus-3-0-sp2-if-https-is-configured-and-tls-1-2-is-not-enabled-1f8540c2-5acf-19b6-6f13-ca85657fce19](url) for details on what this update does.

I think I'll cook up a PR that installs this as a separate update, probably done after the Servicing Stack update, and of course after Windows 8.1 Update 1 is installed.

**Comment by renodr** [comment]
Since this update isn't installed, the LegacyUpdate installer thinks that systems which already have Windows 8.1 Update 1 installed prior to running LegacyUpdate need the whole stack of updates installed, rather than just this one (see the NeedsWin81Update1 function, it would be $2 in this case)

**Comment by renodr** [comment]
I've got a few minor changes to setup.nsi and Download8.nsi that I will open a PR with if they work. On the slipstreamed media from Microsoft with Windows 8.1 Update 1 preinstalled, KB2959977 is in fact not applicable, however LU's installer doesn't pick up on it because the update was superseded in the slipstreamed media, and the file in C:\Windows\servicing\Packages is not present as a result.

Inside of my proposed changes, I removed the check for that update, but still left it in for the installation of Windows 8.1 Update 1 on systems where Windows 8.1 Update 1 actually needs to get installed.

I also discovered a bug where the Servicing Stack Update was not getting offered to systems which already had Windows 8.1 Update 1 installed, so I resolved that as well (just removing a line).

So far I've tested it on just the slipstreamed media, but before I open the PR, I will test it on vanilla Windows 8.1 media (which I have just finished ripping from a DVD) 😃
